aten/src/ATen/core/jit_type.h

```diff
@@ -324,25 +324,50 @@ inline c10::optional<T> merge_primitive(
   return c10::optional<T>{};
 }
 
-// `VaryingShape` tracks if individual dimensions or a rank vary across
-// profiled runs. A *varying* or *dynamic* dimension is expressed as
-// an empty c10::optional in `sizes_`. If a rank is dynamic, the entire
-// `sizes_` becomes the empty optional.
+struct CAFFE2_API ShapeSymbol {
+    ShapeSymbol(int64_t val, bool statik = true): value_(val), statik_(statik) {}
+
+    explicit operator int() const { return value_; }
+    ShapeSymbol(const ShapeSymbol& ss) {
+      statik_ = ss.statik_;
+      value_ = ss.value_;
+    }
+    ShapeSymbol(): value_(-1), statik_(false) {}
+    static ShapeSymbol getInvalidSymbol() {return ShapeSymbol(); }
+    int64_t value_;
+    bool statik_;  
+    bool operator ==(const ShapeSymbol &b) const { return value_ == b.value_ && statik_ == b.statik_; }
+    // dynamic shapes are typically negative
+    bool operator <(const ShapeSymbol &b) const { return static_cast<int>(statik_) < static_cast<int>(b.statik_) || value_ < b.value_; }
+};
+
+template <typename T>
 struct CAFFE2_API VaryingShape {
-  using ListOfOptionalInts = std::vector<c10::optional<int64_t>>;
-  VaryingShape(const std::vector<int64_t>& vec)
-      : VaryingShape(ListOfOptionalInts(vec.begin(), vec.end())) {}
+  using ListOfOptionalElements = std::vector<c10::optional<T>>;
+  VaryingShape(const std::vector<T>& vec)
+      : VaryingShape(ListOfOptionalElements(vec.begin(), vec.end())) {}
 
-  VaryingShape(c10::ArrayRef<int64_t> vec)
-      : VaryingShape(ListOfOptionalInts(vec.begin(), vec.end())){}
+  VaryingShape(c10::ArrayRef<T> vec)
+      : VaryingShape(ListOfOptionalElements(vec.begin(), vec.end())){}
 
   VaryingShape(c10::optional<size_t> size = c10::nullopt) : dims_(c10::nullopt) {
     if (size) {
-      dims_ = ListOfOptionalInts(*size);
+      dims_ = ListOfOptionalElements(*size);
     }
   }
 
-  VaryingShape(ListOfOptionalInts dims)
+  // adds a c-tor VaryingShape<ShapeSymbol>(c10::IntArrayRef)
```

Comment:
I find this confusing, both from a template perspective and from an actual behavior perspective. We don't want to pun integers and shapes but this is extending it through the API.